tests/core/io/test_resource.h

```diff
@@ -109,6 +109,153 @@ TEST_CASE("[Resource] Saving and loading") {
 			loaded_child_resource_text->get_name() == "I'm a child resource",
 			"The loaded child resource name should be equal to the expected value.");
 }
+
+TEST_CASE("[Resource] Thread safety test") {
+	const bool test_simultaneous_load = true;
+	const bool test_cyclic_dependency = false;
```

Comment:
This currently hangs when `true`, but given how pathological the test is (multiple simultaneous requests for resources with cyclic dependencies), it's probably fine to leave it as-is for now.

Conversely, if this actually manages to pass, it would inspire a lot of confidence in the robustness of `ResourceLoader`'s implementation.

Reply:
I'm interested in you elaborating on how pathological the case is. Specifically, do you think the engine has any chance of detecting or preventing the deadlocks?

Reply:
I think it's just a matter of priorities - since deadlocks only seem to be happening with certain improper usages of `ResourceLoader`, it's probably a good idea to merge this PR and [ThreadSanitizer](https://github.com/godotengine/godot/pull/73777) (both are ready now) first so that it's less likely for anything to be accidentally broken in subsequent refactors.

As for methods of preventing cyclic deadlocks, one direction that may be promising (with the obvious caveat that I haven't worked on `ResourceLoader`, and therefore don't have enough experience to rule anything out) is to separate loading into a multithreaded "processing" stage (IO, parsing, decompression, compilation, etc) and a single-threaded "linking" stage.

During "processing", resources would spawn new threads for their immediate children as they're found, but otherwise be self-contained. After "processing" is finished, they would then wait recursively for all children to finish "processing" as well. Finally, the "linking" stage (which can be protected with a global lock for simplicity) would be where all children are attached recursively, and any final action that requires the full tree hierarchy can be performed.

Cycle detection could then be implemented as part of "linking", where it would hopefully be safe from deadlocks.

Reply:
I considered adding some "overview" stage before doing the actual load, that, among other things, would allow to provide more accurate load progress info. I abandoned the idea due to the complexity, but it indeed makes sense to try at some point (now the resource loader is getting more and more stable and it will still have to face new challenges, like resource streaming, maybe for Godot 5.0).

So, for the matter of this PR and the current expected abilities of the resource loader, it may be a better idea to keep the test slimmer and stash the code for cyclic dependencies for an additional test in the future.

Reply:
I'm okay with splitting the tests, but wouldn't it be easier for you if they're already merged into `master` when the time comes, so that they can just be enabled for testing?

Reply:
If you can refactor the current one into two, sharing as much code as reasonable, that sounds good. The other test can be "preprocessed out" with a comment.